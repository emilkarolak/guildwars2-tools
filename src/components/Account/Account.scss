$bg-color: #343A40;
$input-bg-color: #ced4da;

.apikey-bar {
    background-color: $bg-color;
    padding: 1em 0;
    margin-top: 2em;
    input {
        background-color: $input-bg-color;
    }
}

h3 {
    margin: 1em 0;
    button {
        vertical-align: top !important;
    }
}