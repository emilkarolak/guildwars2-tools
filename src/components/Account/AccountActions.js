import LocalCache from '../../helpers/LocalCache';

export const CHANGE_ACCOUNT = 'CHANGE_ACCOUNT';
export const CHANGE_APIKEY = 'CHANGE_APIKEY';
export const REQUEST_ACCOUNT_DATA = 'REQUEST_ACCOUNT_DATA';
export const RECEIVE_ACCOUNT_DATA = 'RECEIVE_ACCOUNT_DATA';

export function changeAccount(account_api_key, accountData) {
  return {
    type: CHANGE_ACCOUNT,
    account_api_key,
    accountData
  };
}

export function changeApikey(account_api_key) {
  return {
    type: CHANGE_APIKEY,
    account_api_key
  };
}

function saveState(accountData) {
  return function() {
    try {
      LocalCache.set('accountState', accountData);
    } catch(error) {}
  }
}

export function loadState() {
  return function(dispatch) {
    return LocalCache.get('accountState').then(({ account_api_key, accountData }) => dispatch(changeAccount(account_api_key, accountData)));
  };
}

function requestAccountData(account_api_key) {
  return {
    type: REQUEST_ACCOUNT_DATA,
    account_api_key
  }
}

function receiveAccountData(account_api_key, json) {
  return {
    type: RECEIVE_ACCOUNT_DATA,
    account_api_key,
    accountData: json,
    receivedAt: Date.now()
  }
}

function fetchAccountData(account_api_key) {
  return dispatch => {
    dispatch(requestAccountData(account_api_key))
    return fetch(`${process.env.REACT_APP_GW2_API_V2}account?access_token=${account_api_key}`)
      .then(response => response.json())
      .then(json => {
          (json.id?dispatch(receiveAccountData(account_api_key, json)):dispatch(receiveAccountData(account_api_key, {})));
          (json.id?dispatch(saveState({account_api_key, accountData: json})):dispatch(saveState({account_api_key, accountData: {}})));
        }
      );
  }
}

function shouldFetchAccountData(state) {
  const { isFetching, didInvalidate } = state.AccountReducer;
  if (!isFetching) {
    return true
  } else if (isFetching) {
    return false
  } else {
    return didInvalidate
  }
}

export function fetchAccountDataIfNeeded(account_api_key) {
  return (dispatch, getState) => {
    if (shouldFetchAccountData(getState())) {
      return dispatch(fetchAccountData(account_api_key))
    }
  }
}