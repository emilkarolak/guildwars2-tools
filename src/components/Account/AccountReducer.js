import {
  CHANGE_ACCOUNT,
  CHANGE_APIKEY,
  REQUEST_ACCOUNT_DATA,
  RECEIVE_ACCOUNT_DATA
} from './AccountActions';

export default function AccountReducer(state = { account_api_key: '', accountData: {}, isFetching: false, didInvalidate: false }, action) {
  switch(action.type) {
    case CHANGE_ACCOUNT:
      return Object.assign({}, state, {
        account_api_key: action.account_api_key,
        accountData: action.accountData
      })
    case CHANGE_APIKEY:
      return Object.assign({}, state, {
        account_api_key: action.account_api_key
      })
    case REQUEST_ACCOUNT_DATA:
      return Object.assign({}, state, {
        isFetching: true,
        didInvalidate: false
      })
    case RECEIVE_ACCOUNT_DATA:
      return Object.assign({}, state, {
        accountData: action.accountData,
        isFetching: false,
        didInvalidate: false,
        lastUpdated: action.receivedAt
      })
    default:
      return state;
  }
}