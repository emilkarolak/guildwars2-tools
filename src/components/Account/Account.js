import React, { useEffect } from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { loadState, changeApikey, fetchAccountDataIfNeeded } from './AccountActions';
import Table from '../../helpers/Table';
import './Account.scss';

const Account = ({ account_api_key, accountData, dispatch }) => {
  const handleChange = (event) => {
    dispatch(changeApikey(event.target.value));
  }
  const handleSubmit = (event) => {
    event.preventDefault();
    dispatch(fetchAccountDataIfNeeded(account_api_key));
  }
  useEffect(() => {
    dispatch(loadState());
  }, [dispatch]);
  return (
    <section className="page-content">
      <h1>
        GuildWars 2 - Account info{accountData?accountData.name?" ("+accountData.name+")":"":""}
      </h1>
      <form onSubmit={handleSubmit} className="row apikey-bar">
        <div className="col-11">
          <div className="row">
            <label htmlFor="apikey" className="col-sm-2 col-form-label">Guild Wars 2 Account API KEY</label>
            <input type="text" id="apikey" className="form-control col-sm-10 apikey-input" value={account_api_key} onChange={handleChange} />
          </div>
        </div>
        <div className="col-1">
          <button type="submit" className="btn btn-primary w-100">Save</button>
        </div>
      </form>
      <div className="row">
        <div className="col-12 mt-3">
          <Table data={accountData} />
        </div>
      </div>
    </section>
  )
};

Account.propTypes = {
  account_api_key: PropTypes.string.isRequired,
  accountData: PropTypes.object.isRequired,
  isFetching: PropTypes.bool.isRequired,
  lastUpdated: PropTypes.number,
  dispatch: PropTypes.func.isRequired
}

function mapStateToProps(state) {
  return state.AccountReducer
}

export default connect(mapStateToProps)(Account)