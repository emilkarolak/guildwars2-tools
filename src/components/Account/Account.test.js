import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Account from './Account';

it('renders without crashing', () => {
  shallow(<Account />);
});

it('renders correctly', () => {
  const tree = renderer
    .create(<Account />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders apikey input bar', () => {
  const wrapper = shallow(<Account />);
  const element = <label htmlFor="apikey" className="col-sm-2 col-form-label">Guild Wars 2 Account API KEY</label>;
  expect(wrapper).toContainReact(element);
});

// it('fetches GW2 account data properly', async () => {
//   const accountComponent = shallow(<Account />);
//   accountComponent.setState({ account_api_key: 'A645F4B5-6F07-B341-9B92-CADF989CAF7F9BE9F9D9-097E-4934-BAA1-3F1B0F5C24AD' });
//   await accountComponent.fetchAccountData();
//   expect(accountComponent.state.accountData).not.toBeNull();
// });

// it('GW2 account data fetch FAILS with an error', async () => {
//   const accountComponent = shallow(<Account />);
//   expect.assertions(1);
//   try {
//     await accountComponent.fetchAccountData();
//   } catch (e) {
//     expect(e).toMatch('error');
//   }
// });