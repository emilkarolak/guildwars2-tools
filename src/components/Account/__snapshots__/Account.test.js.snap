// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="page-content"
>
  <h1>
    GuildWars 2 - Account info
    
  </h1>
  <form
    className="row apikey-bar"
    onSubmit={[Function]}
  >
    <div
      className="col-11"
    >
      <div
        className="row"
      >
        <label
          className="col-sm-2 col-form-label"
          htmlFor="apikey"
        >
          Guild Wars 2 Account API KEY
        </label>
        <input
          className="form-control col-sm-10 apikey-input"
          id="apikey"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-1"
    >
      <button
        className="btn btn-primary w-100"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
  <div
    className="row"
  >
    <div
      className="col-12 mt-3"
    >
      
    </div>
  </div>
</section>
`;
